lom/src/pluginmgr/pluginmgr_common/pluginmgr_helper.go

```diff
@@ -115,63 +165,436 @@ func (plmgr *PluginManager) getPluginMetadata(plgname string) (plugins_common.IP
-                default:
-                    lomcommon.LogError("RecvServerRequest() : Unknown server request type : %v", serverReq.ReqType)
                 }
+            default:
+                lomcommon.LogError("In run() RecvServerRequest : Unknown server request type : %v", serverReq.ReqType)
+            }
+        }
+    }
+    return nil
+}
+
+/*
+ * verify request message
+ */
+func (plmgr *PluginManager) verifyRequestMsg(actionReq *lomipc.ActionRequestData) (plugins_common.Plugin, plugins_common.IPluginMetadata, error) {
+    if actionReq == nil {
+        return nil, nil, lomcommon.LogError("verifyRequestMsg() : nil request data")
+    }
+
+    if actionReq.Action == "" {
+        return nil, nil, lomcommon.LogError("verifyRequestMsg() : empty action")
+    }
+
+    plugin, ok := plmgr.getPlugin(actionReq.Action)
+    if !ok {
+        return nil, nil, lomcommon.LogError("verifyRequestMsg() : Plugin %s not initialized", actionReq.Action)
+    }
+
+    if plugin == nil {
+        return nil, nil, lomcommon.LogError("verifyRequestMsg() : Plugin %s is nil", actionReq.Action)
+    }
+
+    pluginmetadata, ok := plmgr.getPluginMetadata(actionReq.Action)
+    if !ok {
+        return nil, nil, lomcommon.LogError("verifyRequestMsg() : Plugin %s metadata not initialized", actionReq.Action)
+    }
+
+    if pluginmetadata == nil {
+        return nil, nil, lomcommon.LogError("verifyRequestMsg() : Plugin %s metadata is nil", actionReq.Action)
+    }
+
+    pluginStage := pluginmetadata.GetPluginStage()
+    if pluginStage != plugins_common.PluginStageUnknown &&
+        pluginStage != plugins_common.PluginStageLoadingSuccess &&
+        pluginStage != plugins_common.PluginStageRequestSuccess {
+        return nil, nil, lomcommon.LogError("verifyRequestMsg() : Unable to process request for Plugin %s. Reason : %s",
+            actionReq.Action, plugins_common.GetPluginStageToString(pluginStage))
+    }
+
+    return plugin, pluginmetadata, nil
+}
+
+/*
+ * handleRequest : Handle request from server
+ * Basic validation on ActionRequestData is done. Detail validation job is responsibility of plugin
+ */
+func (plmgr *PluginManager) handleRequest(actionReq *lomipc.ActionRequestData) error {
+    // Verify request message and get plugin object for the request action
+    plugin, pluginmetadata, err := plmgr.verifyRequestMsg(actionReq)
+    if err != nil {
+        return err
+    }
+
+    lomcommon.LogInfo(
+        "In handleRequest(): Processing action request for plugin:%s, timeout:%d InstanceId:%s AnomalyInstanceId:%s AnomalyKey:%s",
+        actionReq.Action,
+        actionReq.Timeout,
+        actionReq.InstanceId,
+        actionReq.AnomalyInstanceId,
+        actionReq.AnomalyKey,
+    )
+
+    // create a channel for receiving the responses from plugin
+    respChan := make(chan *lomipc.ActionResponseData, 1)
+
+    // create a channel for receiving the heartbeats from plugin
+    hbChan := make(chan plugins_common.PluginHeartBeat, 1)
```

Comment:
If the size is set to 1, after a heartbeat is written, the execution in plugin continues only after it is read by plugin manager. Should we make it bigger size ?

Reply:
Unless plugin manager is buggy , its not blocking. When HB is received, plugin mgr immediately reads from channel. Also heartbeats are periodic, it shouln't be an issue.